packages/api/internal/sandbox/store/memory/reservation.go

```diff
@@ -10,97 +11,66 @@ import (
 	"github.com/e2b-dev/infra/packages/shared/pkg/utils"
 )
 
-type Reservation struct {
-	sandboxID   string
-	team        uuid.UUID
-	startResult *utils.SetOnce[sandbox.Sandbox]
-}
-
-type ReservationCache struct {
-	reservations *smap.Map[*Reservation]
+type sandboxReservation struct {
+	start *utils.SetOnce[sandbox.Sandbox]
 }
 
-func NewReservationCache() *ReservationCache {
-	return &ReservationCache{
-		reservations: smap.New[*Reservation](),
+func newSandboxReservation(start *utils.SetOnce[sandbox.Sandbox]) *sandboxReservation {
+	return &sandboxReservation{
+		start: start,
 	}
 }
 
-func (r *ReservationCache) insertIfAbsent(sandboxID string, team uuid.UUID, startResult *utils.SetOnce[sandbox.Sandbox]) bool {
-	return r.reservations.InsertIfAbsent(sandboxID, &Reservation{
-		team:        team,
-		sandboxID:   sandboxID,
-		startResult: startResult,
-	})
+type TeamSandboxes struct {
+	teamID    string
+	sandboxes map[string]*sandboxReservation
 }
 
-func (r *ReservationCache) release(sandboxID string) {
-	r.reservations.Remove(sandboxID)
+type ReservationStorage struct {
+	reservations *smap.Map[*TeamSandboxes]
 }
 
-func (r *ReservationCache) list(teamID uuid.UUID) (reservations []*Reservation) {
-	for _, item := range r.reservations.Items() {
-		currentTeamID := item.team
-
-		if currentTeamID == teamID {
-			reservations = append(reservations, item)
-		}
+func NewReservationStorage() *ReservationStorage {
+	return &ReservationStorage{
+		reservations: smap.New[*TeamSandboxes](),
 	}
-
-	return reservations
 }
 
-func (s *Store) list(teamID uuid.UUID) (sandboxIDs []string) {
-	for _, value := range s.items.Items() {
-		currentTeamID := value.TeamID()
-
-		if currentTeamID == teamID {
-			sandboxIDs = append(sandboxIDs, value.SandboxID())
+func (s *ReservationStorage) Reserve(teamID, sandboxID string, limit int64) (finishStart func(sandbox.Sandbox, error), waitForStart func(ctx context.Context) (sandbox.Sandbox, error), err error) {
+	alreadyPresent := false
+	limitExceeded := false
+	var startResult *utils.SetOnce[sandbox.Sandbox]
+
+	s.reservations.Upsert(teamID, nil, func(exist bool, teamSandboxes, _ *TeamSandboxes) *TeamSandboxes {
```

Comment:
Is upsert locked properly when team does not have any sandboxes (missing entry in map)?